### PR TITLE
API and docs improvements

### PR DESCRIPTION
**Pull request by Kriechi**

### CODE REVIEW

File: wsproto/connection.py [hunk 1]
```diff
@@ -186,36 +160,51 @@ def send_data(self, payload, final=True):
...
+        """
         self._outgoing += self._proto.close(code, reason)
         self._state = ConnectionState.CLOSING
 
     @property
     def closed(self):
         return self._state is ConnectionState.CLOSED
 
-    def bytes_to_send(self, amount=None):
+    def bytes_to_send(self, amt=None):
```
Comment (Lukasa) [inline wsproto/connection.py]: Do we need to worry about this slightly-breaking change?

Comment (jeamland) [inline wsproto/connection.py]: I don't like this rename. Explicit is better especially when it's only a few characters.

Comment (Kriechi) [inline wsproto/connection.py]: wsproto is still 0.x.y - so I figured why not.
Again, this is stupidly copying h2 API - no deeper meaning than using the same argument name.

Comment (Kriechi) [reply wsproto/connection.py]: @Lukasa if I revert this here, should I send a PR to h2 to rename it to `amount` as well?

Comment (Lukasa) [reply wsproto/connection.py]: Sure, why not. =) Won't land until we decide to do a 4.0 though. ;)

File: wsproto/connection.py [hunk 2]
```diff
@@ -186,6 +160,13 @@ def send_data(self, payload, final=True):
         self._outgoing += self._proto.send_data(payload, final)
 
     def close(self, code=CloseReason.NORMAL_CLOSURE, reason=None):
+        """
+        Initiate the close handshake by sending a CLOSE control message.
+
+        A clean teardown requires a CLOSE control messages from the other
+        endpoint before the underlying TCP connection can be closed, see
+        :class:`ConnectionClosed <wsproto.events.ConnectionClosed>`.
```
Comment (njsmith) [inline wsproto/connection.py]: You can write this more compactly as
```
:class:`~wsproto.events.ConnectionClosed`
```
([see](http://www.sphinx-doc.org/en/stable/markup/inline.html#cross-referencing-syntax))

File: wsproto/connection.py [hunk 3]
```diff
@@ -195,27 +176,35 @@ def closed(self):
...
-
         return data
 
     def receive_bytes(self, data):
         """
-        Pass some received bytes to the connection for processing.
+        Pass some received data to the connection for handling.
+
+        A list of events that the remote peer triggered by sending this data can
+        be retrieved with :meth:`events <wsproto.connection.WSConnection.events>`.
```
Comment (njsmith) [inline wsproto/connection.py]: Another good place to use sphinx's magic `~`

File: wsproto/utilities.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
+# -*- coding: utf-8 -*-
+"""
+wsproto/utilities
+~~~~~~~~~~~~~~~~~
+
+Utility functions that do not belong in a separate module.
+"""
```
Comment (njsmith) [inline wsproto/utilities.py]: Not sure what the point is of splitting these out – they're not generic utilities that are useful in multiple places, but rather specific to HTTP header parsing, and thus are something that only `wsproto/connection.py` cares about.

File: wsproto/events.py [hunk 5]
```diff
@@ -1,13 +1,20 @@
...
 Events that result from processing data on a WebSocket connection.
 """
 
 
-class ConnectionRequested(object):
+class Event(object):
+    """
+    Base class for wsproto events.
+    """
+    pass
```
Comment (njsmith) [inline wsproto/events.py]: This new empty base class seems kind of pointless to me, but I guess it's harmless too... do you have a use case in mind?

Comment (Kriechi) [reply wsproto/events.py]: I probably had one back then...
hyper-h2 also uses one - seemed like a nice thing todo.

File: wsproto/connection.py (hunk 3 above)
Comment (Kriechi) [reply wsproto/connection.py]: 👍

File: wsproto/connection.py (hunk 2 above)
Comment (Kriechi) [reply wsproto/connection.py]: 👍

File: wsproto/utilities.py (hunk 4 above)
Comment (Kriechi) [reply wsproto/utilities.py]: True.
These functions could be useful elsewhere - but mostly a personal preference what to do with such little helpers. Again, I introduced this when I blindly copied hyper-h2...

**Comment by Kriechi** [comment]
Are there any major issues with this PR?
Should I split it again to make review easier?

If you feel strongly about the empty base class for events, I can revert the change. But IMHO it's a nice thing to do and makes it more streamlined with e.g., hyper-h2 code/API style.

**Comment by smurfix** [comment]
Usecase: You have a common event queue for multiple subsystems and need a way for your dispatcher to figure out which subsystem-specific method to call. That's a lot easier with a common base class, even if it's empty.

**Pull request merged**